File: src/main/java/leetcode/_20/ValidParentheses.java
package leetcode._20;

public class ValidParentheses {

    public boolean isValid(String s) {
        var stack = new char[s.length()];
        var top = 0;

        for (var c : s.toCharArray()) {
            switch (c) {
                case '{', '[', '(' -> stack[top++] = c;
                case '}' -> {
                    if (top == 0 || stack[--top] != '{')
                        return false;
                }
                case ']' -> {
                    if (top == 0 || stack[--top] != '[')
                        return false;
                }
                case ')' -> {
                    if (top == 0 || stack[--top] != '(')
                        return false;
                }
            }
        }

        return top == 0;
    }

}

File: src/main/java/leetcode/_20/valid_parentheses.go
package main

import "fmt"

func isValid(s string) bool {

	stack := make([]byte, len(s))
	top := 0

	for i := 0; i < len(s); i++ {
		switch s[i] {
		case '{', '[', '(':
			stack[top] = s[i]
			top++
			break
		case '}':
			if top == 0 || stack[top-1] != '{' {
				return false
			}
			top--
			break
		case ']':
			if top == 0 || stack[top-1] != '[' {
				return false
			}
			top--
			break
		case ')':
			if top == 0 || stack[top-1] != '(' {
				return false
			}
			top--
			break
		}
	}

	return top == 0
}

func main() {

	s := "(]"
	fmt.Println(isValid(s))
}
